usrtools/txgen/app/txgen.c

```diff
@@ -136,6 +145,77 @@ txgen_fill_pattern(uint8_t *p, uint32_t len, uint32_t type, char *user)
     }
 }
 
+static __inline__ tstamp_t *
+txgen_tstamp_pointer_tx(port_info_t *info, pktmbuf_t *m)
+{
+    tstamp_t *tstamp;
+    uint8_t *p;
+
+    p = pktmbuf_mtod(m, char *);
+
+    p += sizeof(struct cne_ether_hdr);
+
+    p += (info->pkt.ethType == CNE_ETHER_TYPE_IPV4) ? sizeof(struct cne_ipv4_hdr)
+                                                    : sizeof(struct cne_ipv6_hdr);
+
+    p += (info->pkt.ipProto == IPPROTO_UDP) ? sizeof(struct cne_udp_hdr)
+                                            : sizeof(struct cne_tcp_hdr);
+
+    /* Force pointer to be aligned correctly */
+    p = CNE_PTR_ALIGN_CEIL(p, sizeof(uint64_t));
+
+    tstamp = (tstamp_t *)p;
+
+    return tstamp;
+}
+
+static __inline__ tstamp_t *
+txgen_tstamp_pointer_rx(port_info_t *info, pktmbuf_t *m)
+{
+    tstamp_t *tstamp;
+    uint8_t *p;
+
+    p = pktmbuf_mtod_offset(m, char *, sizeof(pktmbuf_t));
```

Comment:
On the rx buffer side I've had to compensate for __get_mbuf_rx_aligned() subtracting the buf_headroom to the offset for the buffer.

If I do pktmbuf_mtod(m, char *); and hexdump the packet data it's all 0s... I'm wondering if you have encountered something similar with cnet?

```
 xskdev_buf_set_data(xi, *bufs, offset - xi->buf_mgmt.buf_headroom);
```
https://github.com/CloudNativeDataPlane/cndp/blob/main/lib/core/xskdev/xskdev.c#L307

Reply:
This does not appear to be right. The offset is the location of the packet data and subtracting buf_headroom seems wrong. I do not see this problem in CNET as nothing would work.

Reply:
this is the only thing that prints out actual packet data at that offset. Vs all zeros from the 0 offset...

> I do not see this problem in CNET as nothing would work.

That's what I thought... so, I'm not sure if there's a setup issue with txgen that I missed... or if something has changed under the hood of xdpdev in the newer AF_XDP APIs... OR it's just something I'm missing.

The idea for the time stamping is pretty straight forward. On the TX side insert the timestamp at the start of the data part of the packet. On RX side, read the timestamp and calculate the latency.

if I just try to setup p as:
```
p = pktmbuf_mtod(m, char *);
```
and then hexdump the contents at that location - I see 0s for sizeof(pktmuf_t)...

If I offset as shown I'm reading the timestamp from the right place... so something is a miss...

Reply:
The latency timestamp needs to be placed at data_offset + sizeof(L2 header) + sizeof(L3 header) + sizeof(L4 header). As the data_offset should point to start of the packet unless it has been updated. Adding sizeof(pktmbuf_t) is only moving the pointer by 64 bytes.

Reply:
> The latency timestamp needs to be placed at data_offset + sizeof(L2 header) + sizeof(L3 header) + sizeof(L4 header). As the data_offset should point to start of the packet unless it has been updated. Adding sizeof(pktmbuf_t) is only moving the pointer by 64 bytes.

That is where I'm putting it (and reading it from). At the `payload` part of the packet.

If I just look at the incoming buffer and print out the contents from data_offset there's a bunch of zeros and not the L2 header as I was expecting.

Reply:
One of the patches we added was to move the umem.frame_headroom by sizeof(pktmbuf_t) to account for VPP needing to add its sizeof(vlib_buffer_t) to skip overwriting the pktmbuf_t header. This change was in the VPP example plugin code.

Now the umem.frame_headroom is defaulted to sizeof(pktmbuf_t) as the offset. When the frame_headroom was zero all RX data offsets were placed at 256 bytes from the beginning of the UMEM buffer. Now the packet data on RX will start at 256 + sizeof(pktmbuf_t). Need to verify this is the case as I do not remember exactly.

For txgen it should be not updating the data_offset from the xskdev driver. Can you tell the exact offset from the beginning of the UMEM buffer the packet data is being placed?

Reply:
Did you have any luck on finding the problem here?

Reply:
Will try to add more details in the morning. didn't get a chance to debug more today

Reply:
I grab your branch and put a hexdump in the txgen_tstamp_pointer_rx() routine and I am seeing L2 headers. I am not sending true latency packet, but regular UDP packets. I will look the transmit side real quick.

Reply:
I think line 178 in txgen.c inside the txgen_tstamp_pointer_rx()

p = pktmbuf_mtod_offset(m, char *, sizeof(pktmbuf_t));
should be
p = pktmbuf_mtod(m, char *);

A the offset zero is where the L2 header starts.

Reply:
> cne_hexdump(NULL, pktmbuf_mtod(pkt, char *), 16);

Hey Keith
I did a quick test this morning, setup a UDP packet and sent it through. I added some instrumentation to _process_tstamp() below:

```
static __inline__ void
_process_tstamp(port_info_t *info, pktmbuf_t *pkt)
{
    uint64_t lat, jitter;
    tstamp_t *tstamp;

	cne_printf("pktmbuf_mtod(pkt, char *)\n");
	cne_hexdump(NULL, "pktmbuf_mtod(pkt, char *)", pktmbuf_mtod(pkt, char *), 16);
	cne_printf("pktmbuf_mtod_offset(pkt, char *, sizeof(pktmbuf_t)\n");
	cne_hexdump(NULL, "pktmbuf_mtod_offset(pkt, char *, sizeof(pktmbuf_t)", pktmbuf_mtod_offset(pkt, char *, sizeof(pktmbuf_t)), 16);

    tstamp = txgen_tstamp_pointer_rx(info, pkt);
    if (tstamp->magic == TSTAMP_MAGIC) {
        lat                    = (cne_rdtsc_precise() - tstamp->timestamp);
        latsamp_stats_t *stats = &info->latsamp_stats;
        uint64_t now           = cne_rdtsc_precise();

        info->avg_latency += lat;
        if (lat > info->prev_latency)
            jitter = lat - info->prev_latency;
        else
            jitter = info->prev_latency - lat;
        if (lat > info->max_latency)
            info->max_latency = lat;
        if (jitter > info->jitter_threshold_clks)
            info->jitter_count++;
        info->prev_latency = lat;

        stats->pkt_counter++;
        if (stats->next == 0 || now >= stats->next) {
            if (stats->idx < stats->num_samples) {
                stats->data[stats->idx] =
                    (lat * Billion) / cne_get_timer_hz(); /* Do we want to keep it as cycles? */
                stats->idx++;
            }

            /* Calculate next sampling point TODO: Use poisson */
            if (info->latsamp_type == LATSAMPLER_POISSON) {
                double next_possion_time_ns = next_poisson_time(info->latsamp_rate);
                stats->next =
                    now + next_possion_time_ns * (double)cne_get_timer_hz();        // Time based
            } else {        // LATSAMPLER_SIMPLE or LATSAMPLER_UNSPEC
                stats->next = now + cne_get_timer_hz() / info->latsamp_rate;        // Time based
            }
        }
    } else {
        info->magic_errors++;
    }

    info->latency_nb_pkts++;
}

```

I only see the L2 headers when I'm using pktmbuf_mtod_offset(pkt, char *, sizeof(pktmbuf_t)

```
TXGen:/> pktmbuf_mtod(pkt, char *)
 at [0x7f9caec000c0], len=16
00000000: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | ................
pktmbuf_mtod_offset(pkt, char *, sizeof(pktmbuf_t)
 at [0x7f9caec00100], len=16
00000000: EC F4 BB C0 B6 2A EC F4 BB C0 B6 28 08 00 45 00 | .....*.....(..E.
```

What version of libbpf are you using?

I'm using libxdp and I'm wondering if something changed under the hood. I can try to use the same version of libbpf as you to see if I can duplicate the behaviour you are seeing (L2 headers at offset 0)?

Reply:
these are my bpf and xdp lib versions:

```
Run-time dependency libxdp found: YES 1.2.5
Run-time dependency libbpf found: YES 0.8.0
```

Reply:
I am using Run-time dependency libbpf found: YES 0.6.0, but I need to move to libxdp. Any libbpf above 6.1 will cause a set of warning as the deprecated the xsk APIs.

Reply:
let me bounce back to 0.6.0 and see what the behaviour is there with the offset weirdness...

Reply:
so I tried it with 0.6.0 and I see the L2 header with  pktmbuf_mtod(pkt, char *)... so it looks like something did change under the hood. I don't know if this is a bug we want to flag to Magnus, but for now I will submit a patch that will fix the issue for CNDP applications using libxdp 

```
TXGen:/> pktmbuf_mtod(pkt, char *)
 at [0x7f833ec01900], len=16
00000000: 02 A7 A2 BC 51 30 EC F4 BB C0 B6 2A 08 00 45 00 | ....Q0.....*..E.
pktmbuf_mtod_offset(pkt, char *, sizeof(pktmbuf_t)
 at [0x7f833ec01940], len=16
00000000: 61 62 63 64 65 66 67 68 69 6A 6B 6C 6D 6E 6F 70 | abcdefghijklmnop

```

Reply:
Because the APIs in libbpf 0.8.0 are deprecated it maybe they did not go back and fix something they added in libxdp.

Reply:
We even have this noted in the INSTALL.md file and it does fix the build issue. :-(